homeassistant/components/xiaomi_miio/gateway.py

```diff
@@ -33,16 +44,48 @@ def gateway_info(self):
     async def async_connect_gateway(self, host, token):
         """Connect to the Xiaomi Gateway."""
         _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+
+        use_cloud = self._config_entry.options.get(CONF_CLOUD_SUBDEVICES, False)
+        cloud_username = self._config_entry.data.get(CONF_CLOUD_USERNAME)
+        cloud_password = self._config_entry.data.get(CONF_CLOUD_PASSWORD)
+        cloud_country = self._config_entry.data.get(CONF_CLOUD_COUNTRY)
+
         try:
             self._gateway_device = gateway.Gateway(host, token)
             # get the gateway info
             self._gateway_info = await self._hass.async_add_executor_job(
                 self._gateway_device.info
             )
+
             # get the connected sub devices
-            await self._hass.async_add_executor_job(
-                self._gateway_device.discover_devices
-            )
+            if use_cloud or self._gateway_info.model == GATEWAY_MODEL_EU:
+                if (
+                    cloud_username is None
+                    or cloud_password is None
+                    or cloud_country is None
+                ):
+                    raise ConfigEntryAuthFailed(
+                        "Missing cloud credentials in Xiaomi Miio configuration"
+                    )
+
+                # use miio-cloud
+                miio_cloud = MiCloud(cloud_username, cloud_password)
+                if not await self._hass.async_add_executor_job(miio_cloud.login):
```

Comment:
Can we combine the sync calls inside a function that we schedule once on the executor? Context switches are expensive.

Reply:
done